include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -840,6 +840,173 @@ class __static_monotonic_dispatcher<::std::integer_sequence<::std::uint16_t, _X,
     }
 };
 
+struct __scalar_load_op
+{
+    oneapi::dpl::__internal::__pstl_assign __assigner;
+    template <typename _IdxType1, typename _IdxType2, typename _SourceAcc, typename _DestAcc>
+    void
+    operator()(_IdxType1 __idx_source, _IdxType2 __idx_dest, _SourceAcc __source_acc, _DestAcc __dest_acc) const
+    {
+        __assigner(__source_acc[__idx_source], __dest_acc[__idx_dest]);
+    }
+};
+
+template <std::uint8_t __vec_size>
+struct __vector_load
```

Comment:
1) Why do we need a such struct? It looks that we need just one `vec_load `function...
2) Whether can we keep just one impl, like ?
```
        std::uint8_t __elements = std::min(std::size_t{__vec_size}, std::size_t{__full_range_size - __start_idx});
        _ONEDPL_PRAGMA_UNROLL
        for (std::uint8_t __i = 0; __i < __elements; ++__i)
            __load_op(__start_idx + __i, __i, __rngs...);
```

Reply:
The unrolling only works with compile time knowledge of the loop bounds.

Reply:
Yes, as Dan says `__elements` needs to be a compile-time variable here. Otherwise, vector instructions cannot be generated.

Reply:
Just one other follow-up on point 1. With the recent changes, if `__vector_load` were to be a function, our uses would be `__vector_load<__base_t::__preferred_vector_size>(...)` each use instead of creating the object once, passing the template once, and then calling `__vec_load(...)` as many times as needed.

Reply:
> The unrolling / vectorization only works with compile time knowledge of the loop bounds.

It is so big stone into compiler side :-)
It looks we are coming back to the  first С compiler, written approximately  50 yeas ago...)

Reply:
Yes, this is a big cause of complexity for our sycl implementations, especially because unrolling / vectorization can make a BIG difference in performance.  

I've long thought that we could have some generic loop wrappers which automatically do this for the compiler by unrolling to some fixed size based on the element types.  This is effectively what this PR does.  There may be some ways we can extend it for more general usage as a tool in other patterns.